### ISSUE TITLE
NPM jointjs type definitions - PaperOptions do not contain model

### ISSUE THREAD

**Issue by denisemauldin** [clientIO/joint#582]
Using Angular 2.4.0, tsc 2.3.1, I'm trying to add JointJS

```
npm install jointjs --save
npm install backbone@1.3.3 --save
npm install lodash@3.10.1 --save
npm install jquery@3.1.1 --save
```

Using this example:
http://resources.jointjs.com/demos/joint/demo/links/src/links.js

Just copying the first two links code and adding it to a method in my Angular 2 component:

    import { Component, OnInit } from '@angular/core';
    
    import * as joint from 'jointjs/dist/joint.js';
    import * as $ from 'jquery';
    
    @Component({
      selector: 'app-root',
      templateUrl: './app.component.html',
      styleUrls: ['./app.component.css']
    })
    export class AppComponent implements OnInit {
      graph: any;
      paper: any;
    
      ngOnInit(){
        this.createGraph();
      }
    
      private createGraph() {
        var graph = new joint.dia.Graph();
    var paper = new joint.dia.Paper({
        el: $('#paper'),
        width: 800,
        height: 600,
        model: graph,
        gridSize: 1
    });
    
    var link = new joint.dia.Link({
        source: { x: 10, y: 20 },
        target: { x: 350, y: 20 },
        attrs: {}
    });
    
    /*
    $rappid_green: #31d0c6;
    $rappid_purple: #7c68fc;
    $rappid_orange: #fe854f;
    $rappid_orange2: #feb663;
    $rappid_white: #f6f6f6;
    $rappid_grey1: #222138;
    $rappid_grey2: #33334e;
    $rappid_grey3: #4b4a67;
    $rappid_orange4: #3c4260;
    $rappid_grey5: #6a6c8a;
    $rappid_grey6: #c6c7e2;
    $payne_grey: #3c4260;
    */
    
    link.attr({
        '.connection': { stroke: '#222138' },
        '.marker-source': { fill: '#31d0c6', stroke: 'none', d: 'M 10 0 L 0 5 L 10 10 z' },
        '.marker-target': { fill: '#fe854f', stroke: '#7c68fc', d: 'M 10 0 L 0 5 L 10 10 z' }
    });
    
    var link2 = new joint.dia.Link({
        source: { x: 10, y: 80 },
        target: { x: 350, y: 80 },
        attrs: {}
    });
    
    link2.attr({
        '.connection': { stroke: '#fe854f', 'stroke-width': 4 },
        '.marker-source': { stroke: '#fe854f', fill: '#fe854f', d: 'M 10 0 L 0 5 L 10 10 z' },
        '.marker-target': { stroke: '#fe854f', fill: '#fe854f', d: 'M 10 0 L 0 5 L 10 10 z' }
    });
         graph.addCell([link, link2]);
      }
    
    }

Results in a TS Error:

> ERROR in src/app/app.component.ts (21,99): Argument of type '{ el:
> JQuery; width: number; height: number; gridSize: number; model: any;
> perpendicularLinks: tr...' is not assignable to parameter of type
> 'PaperOptions'.   Object literal may only specify known properties,
> and 'model' does not exist in type 'PaperOptions'.)

Indeed, if you open `node_modules/jointjs/types/joint.d.ts` the `interface PaperOptions` does not contain `model`.  How do I fix this?

This is a small repo that also displays the same error:
https://github.com/CoconutTheSlayer/jointangular2

**Comment by denisemauldin** [comment]
Ok.  If I do `npm install -D @types/jointjs` then it stops throwing the `'model' does not exist in type 'PaperOptions'` error, but it throws a bunch of other errors:

```
node_modules/jointjs/dist/joint.d.ts(49,21): error TS7006: Parameter 'c' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(49,24): error TS7006: Parameter 'a' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(49,27): error TS7006: Parameter 'b' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(101,23): error TS7006: Parameter 'distance' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(101,33): error TS7006: Parameter 'angle' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(101,40): error TS7006: Parameter 'origin' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(112,9): error TS7010: 'changeInAngle', which lacks return-type annotation, implicitly has an 'any' return type.
node_modules/jointjs/dist/joint.d.ts(112,23): error TS7006: Parameter 'dx' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(112,27): error TS7006: Parameter 'dy' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(112,31): error TS7006: Parameter 'ref' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(125,14): error TS7006: Parameter 'ref' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(125,19): error TS7006: Parameter 'distance' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(155,21): error TS7006: Parameter 'x' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(155,24): error TS7006: Parameter 'y' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(155,27): error TS7006: Parameter 'w' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(155,30): error TS7006: Parameter 'h' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(240,19): error TS7006: Parameter 'x' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(240,21): error TS7006: Parameter 'y' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(240,23): error TS7006: Parameter 'w' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(240,25): error TS7006: Parameter 'h' implicitly has an 'any' type.
```

**Comment by denisemauldin** [comment]
If I delete the node_modules/jointjs/dist/joint.d.ts file then it builds with no errors.

**Comment by vtalas** [comment]
>Ok. If I do npm install -D @types/jointjs then it stops throwing the 'model' does not exist in type 'PaperOptions' error, but it throws a bunch of other errors:

>node_modules/jointjs/dist/joint.d.ts(49,21): error TS7006: Parameter 'c' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(49,24): error TS7006: Parameter 'a' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(49,27): error TS7006: Parameter 'b' implicitly has an 'any' type.
node_modules/jointjs/dist/joint.d.ts(101,23): error TS7006: Parameter 'distance' implicitly ...

tsc compile option `--noImplicitAny=false` should solve this. 

> Indeed, if you open node_modules/jointjs/types/joint.d.ts the interface PaperOptions does not contain model. How do I fix this?

the `model` property is defined on the parent of the `PaperOptions` : 
`interface PaperOptions extends Backbone.ViewOptions<Graph> { ...`
I'm not sure what could go wrong... 

typescript definition file for jointjs is in the first version, there are still some bugs which will be hopefuly fixed in the following release. Please note the `@types/joinjs` is not needed, the most recent version should always come with the JointJs release (located in the `dist/joint.d.ts`)

**Comment by denisemauldin** [comment]
Well, backbone 1.3.3 doesn't seem to include a `Backbone.ViewOptions`.  There is a `viewOptions` inside of `Backbone.View`.  Maybe ViewOptions is something in a newer version of Backbone?  Maybe ViewOptions is a plugin (https://github.com/rotundasoftware/backbone.viewOptions) that needs to be installed but isn't mentioned on the Joint website? ;)

**Comment by kumilingus** [comment]
Fixed in `v2.0.0`.
